File: frontend/types/commerce.ts
export type ProductType = "TEST" | "SUBSCRIPTION";
export type PaymentStatus = "CREATED" | "PAID" | "FAILED" | "REFUNDED";
export type SubscriptionStatus = "PENDING" | "ACTIVE" | "CANCELLED" | "EXPIRED";

export interface Product {
  id: string;
  name: string;
  description: string | null;
  product_type: ProductType;
  price_paise: number;
  price_inr: number;
  currency: string;
  question_limit: number | null;
  duration_seconds: number | null;
  billing_interval_days: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateOrderResponse {
  order_id: string;
  amount_paise: number;
  amount_inr: number;
  currency: string;
  product_id: string;
  product_name: string;
  product_type: ProductType;
  razorpay_key_id: string | null;
  is_mock_mode: boolean;
}

export interface PaymentVerificationResponse {
  success: boolean;
  payment_id: string;
  order_id: string;
  status: PaymentStatus;
  message: string;
  entitlement_id: string | null;
  subscription_id: string | null;
}

export interface MembershipStatus {
  user_id: string;
  email: string;
  full_name: string | null;
  role: string;
  is_subscribed: boolean;
  subscription_status: SubscriptionStatus | null;
  subscription_plan: string | null;
  subscription_end_date: string | null;
  available_test_credits: number;
  free_test_consumed: boolean;
}

export interface PaymentRecord {
  id: string;
  product_name: string;
  product_type: ProductType;
  amount_inr: number;
  currency: string;
  status: PaymentStatus;
  provider: string;
  provider_order_id: string | null;
  provider_payment_id: string | null;
  created_at: string;
}

export interface UserAttemptRecord {
  id: string;
  test_id: string;
  test_name: string;
  status: string;
  score: number | null;
  total_questions: number;
  correct_count: number | null;
  wrong_count: number | null;
  skipped_count: number | null;
  accuracy: number | null;
  time_taken_seconds: number | null;
  started_at: string;
  submitted_at: string | null;
}

export interface TopicWeakness {
  category_name: string;
  topic_name: string;
  total_questions: number;
  correct_questions: number;
  accuracy: number;
}

export interface StudentDashboardData {
  user_name: string;
  email: string;
  role: string;
  is_subscribed: boolean;
  subscription_plan: string | null;
  subscription_end_date: string | null;
  available_test_credits: number;
  free_test_consumed: boolean;
  tests_completed: number;
  questions_attempted: number;
  overall_accuracy: number;
  average_time_per_question_seconds: number;
  best_score: number | null;
  best_score_total: number | null;
  best_accuracy: number | null;
  streak_days: number;
  weak_topics: TopicWeakness[];
  recent_attempts: UserAttemptRecord[];
}

export interface RecentRegistration {
  id: string;
  email: string;
  full_name: string | null;
  role: string;
  is_active: boolean;
  created_at: string;
}

export interface RecentPaymentRecord {
  id: string;
  user_email: string;
  amount_inr: number;
  status: PaymentStatus;
  product_name: string;
  created_at: string;
}

export interface PopularTestRecord {
  test_id: string;
  test_name: string;
  attempt_count: number;
  average_accuracy: number;
}

export interface AdminAnalyticsData {
  total_users: number;
  total_tests_taken: number;
  total_questions_in_bank: number;
  total_revenue_inr: number;
  paid_purchases_count: number;
  active_subscriptions_count: number;
  recent_registrations: RecentRegistration[];
  recent_payments: RecentPaymentRecord[];
  popular_tests: PopularTestRecord[];
}

File: backend/app/schemas/commerce.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.enums import PaymentStatus, ProductType, SubscriptionStatus


class ProductBase(BaseModel):
    name: str = Field(min_length=3, max_length=180)
    description: str | None = None
    product_type: ProductType
    price_paise: int = Field(ge=0)
    currency: str = "INR"
    question_limit: int | None = None
    duration_seconds: int | None = None
    billing_interval_days: int | None = None
    is_active: bool = True


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price_paise: int | None = Field(default=None, ge=0)
    question_limit: int | None = None
    duration_seconds: int | None = None
    billing_interval_days: int | None = None
    is_active: bool | None = None


class ProductRead(ProductBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def price_inr(self) -> int:
        return self.price_paise // 100


class CreateOrderRequest(BaseModel):
    product_id: UUID


class CreateOrderResponse(BaseModel):
    order_id: str
    amount_paise: int
    amount_inr: int
    currency: str
    product_id: UUID
    product_name: str
    product_type: ProductType
    razorpay_key_id: str | None
    is_mock_mode: bool


class VerifyPaymentRequest(BaseModel):
    order_id: str
    payment_id: str
    signature: str


class PaymentVerificationResponse(BaseModel):
    success: bool
    payment_id: UUID
    order_id: str
    status: PaymentStatus
    message: str
    entitlement_id: UUID | None = None
    subscription_id: UUID | None = None


class MembershipStatusResponse(BaseModel):
    user_id: UUID
    email: str
    full_name: str | None
    role: str
    is_subscribed: bool
    subscription_status: SubscriptionStatus | None = None
    subscription_plan: str | None = None
    subscription_end_date: datetime | None = None
    available_test_credits: int = 0
    free_test_consumed: bool = False


class PaymentHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    product_name: str
    product_type: ProductType
    amount_inr: int
    currency: str
    status: PaymentStatus
    provider: str
    provider_order_id: str | None
    provider_payment_id: str | None
    created_at: datetime


class UserAttemptHistoryRead(BaseModel):
    id: UUID
    test_id: UUID
    test_name: str
    status: str
    score: int | None
    total_questions: int
    correct_count: int | None
    wrong_count: int | None
    skipped_count: int | None
    accuracy: float | None
    time_taken_seconds: int | None
    started_at: datetime
    submitted_at: datetime | None


class TopicWeakness(BaseModel):
    category_name: str
    topic_name: str
    total_questions: int
    correct_questions: int
    accuracy: float


class StudentDashboardResponse(BaseModel):
    user_name: str
    email: str
    role: str
    is_subscribed: bool
    subscription_plan: str | None = None
    subscription_end_date: datetime | None = None
    available_test_credits: int = 0
    free_test_consumed: bool = False
    tests_completed: int = 0
    questions_attempted: int = 0
    overall_accuracy: float = 0.0
    average_time_per_question_seconds: float = 0.0
    best_score: int | None = None
    best_score_total: int | None = None
    best_accuracy: float | None = None
    streak_days: int = 0
    weak_topics: list[TopicWeakness] = []
    recent_attempts: list[UserAttemptHistoryRead] = []

